Exercise 9

JoeOsborne22 opened the pull request:
Updated APP to work with Mongo instead of TRELLO
All test packs updated to reflect changes
All Tests pass in CI Pipeline on Github 
Disabled deployment to Heroku due to issues encountered and possible issues related to recent attack

JoeOsborne22 committed:
exercise-9
	- First iteration of changes towards swapping integration of
	  TRELLO with MongoDB, not integration tests & CI/CD pipeline
	  still outstanding

JoeOsborne22 committed:
exercise-9
	- Update tests to correspond with migration to mongo

JoeOsborne22 committed:
exercise-9
	- Update git actions file

JoeOsborne22 committed:
exercise-9
	- revert pipeline cfg back to original

JoeOsborne22 committed:
exercise -9
	- Update integration tests further and update CI_Pipeline to run
	  e2e tests as expected

JoeOsborne22 committed:
exercise-9
	- Update README

JoeOsborne22 committed:
exercise-9
	- Update CI Pipeline.yml

JoeOsborne22 committed:
exercise-9
	- Github action to deploy to heroku has an issue, attemting to
	  use heroku to deply instead

hugh-emerson commented on todo_app/ToDoItem.py:
```diff
@@ -16,14 +20,17 @@ def __init__(self, lists, json):
-            cardDate=(datetime.fromisoformat(self.due[:-1])).date()
+            try:cardDate=(datetime.fromisoformat(self.due[:-1])).date()
+            except :cardDate=datetime.strptime(self.due, '%Y-%m-%d').date()
+                #cardDate=datetime.combine(datetime.strptime(self.due, '%Y-%m-%d'), datetime.min.time())
```
Delete commented out code

hugh-emerson commented on todo_app/app.py:
```diff
@@ -1,37 +1,40 @@
     @app.route('/')
     def index():
-        cards=getTrelloCards(trelloBoardID, key, token)
+        cards=getTrelloCards()
```
A lot of references to "Trello" are now out of date.

hugh-emerson commented on todo_app/app.py:
```diff
@@ -42,9 +45,10 @@ def add_item():
         if not card_title == "": 
-            cardDetails={"idList":trelloBaseList,"name":card_title,"pos":"top","desc":card_desc,"due":card_due}
-            updateTrelloCard('', cardDetails, "POST", key, token) 
+            cardDetails={"idList":defaultStatus,"name":card_title,"pos":"top","desc":card_desc,"due":card_due,"dateLastActivity":card_lastActivity}
```
Slightly confusing to stick to calling the status "idList" for legacy reasons instead of naming it "status".

hugh-emerson commented on todo_app/app.py:
```diff
@@ -61,52 +65,35 @@ def sort_item():
-        res=requestJson("GET",url, {})
+    # Used to update an item
+    def updateTrelloCard (newState, method):
+        try: res=actionMap[method](newState)
+        except: res=actionMap[method](newState[0],newState[1])
+        
```
This is a little confusing and likely to lead to bugs.

Is it really neater to wrap `todoCards.update_one`/`todoCards.delete_one` in a single `updateTrelloCard` function when they are such different functions? Before you were wrapping making HTTP requests to an API. So the interface is always HTTP and there's some common behaviour like key, token, parsing the body as JSON. That made sense.

But this function obscures what's happening and is also less convenient to use.
If you just make them separate functions you could have functions that are used like this `deleteTrelloCard(card_id)` instead of `updateTrelloCard({"_id":card_id},"DELETE")`.

hugh-emerson commented on .github/workflows/CI_Pipeline.yml:
```diff
@@ -25,20 +25,19 @@ jobs:
-    steps:
-    - uses: actions/checkout@v2 # Adds a step to checkout the repository code
+    - run: docker run -e MONGO_DB_NAME=${{secrets.MONGO_DB_NAME}} -e MONGO_CONNECT=${{secrets.MONGO_CONNECT}} -e MONGO_COLLECTION_NAME=${{secrets.MONGO_COLLECTION_NAME}} -e MONGO_DEFAULT_STATUS=${{secrets.MONGO_DEFAULT_STATUS}} my-test-image tests_e2e
+#  deploy:
+#    name: Deploy build package to Heroku
+#    needs: build
+#    if: ${{ github.event_name ==  'push' }}
+#    runs-on: ubuntu-latest
+#    steps:
+#    - uses: actions/checkout@v2 # Adds a step to checkout the repository code
```
Why is this commented out?

JoeOsborne22 replied:
Hi Hugh, 

Thanks for the feedback, I had popped a message on slack regarding this and it seems this GitHub action no longer works and when I was searching it seemed to be related to the recent security issue so I went down the route of commenting out for the time being.

hugh-emerson commented on tests/test_integrationToDoApp.py:
```diff
@@ -27,73 +35,36 @@ def test_index_page(mock_get_requests, client):
+def addMockData():
 
+    mongoDB=os.getenv('MONGO_DB_NAME')
+    mongoTable=os.getenv('MONGO_TABLE_NAME')
```
I think you mean MONGO_COLLECTION_NAME instead of MONGO_TABLE_NAME

Your .env.template has both, but currently only TABLE_NAME is used and "COLLECTION" is the correct term.

hugh-emerson commented on .github/workflows/CI_Pipeline.yml:
```diff
@@ -25,20 +25,19 @@ jobs:
-    runs-on: ubuntu-latest
-    steps:
-    - uses: actions/checkout@v2 # Adds a step to checkout the repository code
+    - run: docker run -e MONGO_DB_NAME=${{secrets.MONGO_DB_NAME}} -e MONGO_CONNECT=${{secrets.MONGO_CONNECT}} -e MONGO_COLLECTION_NAME=${{secrets.MONGO_COLLECTION_NAME}} -e MONGO_DEFAULT_STATUS=${{secrets.MONGO_DEFAULT_STATUS}} my-test-image tests_e2e
```
It's overkill to make the db name, collection name, and default status into secrets.

You could just set e.g. `-e MONGO_DB_NAME=e2e_db`

hugh-emerson reviewed (approved):
Looks good except you seem to be avoiding deploying to Heroku. Let us know if you need any help getting it to work there.

I've left some optional suggestions but go ahead and merge when you're ready 👍

JoeOsborne22 committed:
module-9 Update CI Pipeline to try new Github action

JoeOsborne22 committed:
module-9
	- Revert change as deploy to heroku still has issues - this will
	  be resolved in future module

JoeOsborne22 committed:
exercise-9
	- Applied reccomendations from PR

JoeOsborne22 committed:
exercise-9
	- Asset statements to check data within integration tests having
	  some issue

JoeOsborne22 committed:
exercise-9
	- do not deploy due to issues

JoeOsborne22 committed:
exercise-9 attempt heroku deploy once more
